File: sdk-python/wave/distribution.py
"""WAVE SDK - Distribution API. Social simulcasting and scheduled publishing."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class Destination(BaseModel):
    id: str; organization_id: str; name: str; type: str; status: str; auto_start: bool = False; created_at: str; updated_at: str

class DistributionAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/distribution"
    def list_destinations(self, **params: Any) -> dict: return self._client.get(f"{self._base}/destinations", params={k: v for k, v in params.items() if v is not None})
    def get_destination(self, dest_id: str) -> Destination: return Destination(**self._client.get(f"{self._base}/destinations/{dest_id}"))
    def add_destination(self, name: str, type: str, **kwargs: Any) -> Destination: return Destination(**self._client.post(f"{self._base}/destinations", json={"name": name, "type": type, **kwargs}))
    def remove_destination(self, dest_id: str) -> None: self._client.delete(f"{self._base}/destinations/{dest_id}")
    def start_simulcast(self, stream_id: str, destination_ids: list[str]) -> dict: return self._client.post(f"{self._base}/simulcast", json={"stream_id": stream_id, "destination_ids": destination_ids})
    def stop_simulcast(self, stream_id: str) -> dict: return self._client.post(f"{self._base}/simulcast/stop", json={"stream_id": stream_id})
    def get_simulcast_status(self, stream_id: str) -> dict: return self._client.get(f"{self._base}/simulcast/{stream_id}")
    def schedule_post(self, title: str, platforms: list[str], media_url: str, scheduled_at: str, **kwargs: Any) -> dict: return self._client.post(f"{self._base}/posts", json={"title": title, "platforms": platforms, "media_url": media_url, "scheduled_at": scheduled_at, **kwargs})
    def list_scheduled_posts(self, **params: Any) -> dict: return self._client.get(f"{self._base}/posts", params={k: v for k, v in params.items() if v is not None})
    def cancel_scheduled_post(self, post_id: str) -> None: self._client.delete(f"{self._base}/posts/{post_id}")

File: sdk-typescript/packages/distribution/src/distribution.ts
/**
 * WAVE SDK - Distribution API
 *
 * Social media distribution, simulcasting, and scheduled publishing.
 */

import type { WaveClient, PaginationParams, PaginatedResponse, Timestamps } from '@wave-av/core';

export type DestinationType =
  | "youtube"
  | "twitch"
  | "facebook"
  | "linkedin"
  | "twitter"
  | "tiktok"
  | "instagram"
  | "custom_rtmp";
export type DestinationStatus = "connected" | "disconnected" | "streaming" | "error";

export interface Destination extends Timestamps {
  id: string;
  organization_id: string;
  name: string;
  type: DestinationType;
  status: DestinationStatus;
  rtmp_url?: string;
  stream_key_ref?: string;
  platform_channel_id?: string;
  auto_start: boolean;
}
export interface SimulcastSession {
  id: string;
  stream_id: string;
  destinations: SimulcastTarget[];
  status: "active" | "stopped";
  started_at: string;
  stopped_at?: string;
}
export interface SimulcastTarget {
  destination_id: string;
  status: "streaming" | "error" | "pending";
  viewer_count?: number;
  error_message?: string;
}
export interface ScheduledPost extends Timestamps {
  id: string;
  organization_id: string;
  title: string;
  description: string;
  platforms: DestinationType[];
  media_url: string;
  scheduled_at: string;
  status: "scheduled" | "publishing" | "published" | "failed";
  published_urls?: Record<string, string>;
}
export interface AddDestinationRequest {
  name: string;
  type: DestinationType;
  rtmp_url?: string;
  stream_key_ref?: string;
  platform_channel_id?: string;
  auto_start?: boolean;
}
export interface ListDestinationsParams extends PaginationParams {
  type?: DestinationType;
  status?: DestinationStatus;
}

/**
 * Social media simulcasting and scheduled content publishing.
 *
 * @example
 * ```typescript
 * await wave.distribution.addDestination({ name: "YouTube", type: "youtube", auto_start: true });
 * await wave.distribution.startSimulcast(streamId, [destId1, destId2]);
 * ```
 */
export class DistributionAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/distribution";
  constructor(client: WaveClient) {
    this.client = client;
  }

  async listDestinations(params?: ListDestinationsParams): Promise<PaginatedResponse<Destination>> {
    return this.client.get<PaginatedResponse<Destination>>(`${this.basePath}/destinations`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async getDestination(destId: string): Promise<Destination> {
    return this.client.get<Destination>(`${this.basePath}/destinations/${destId}`);
  }
  async addDestination(request: AddDestinationRequest): Promise<Destination> {
    return this.client.post<Destination>(`${this.basePath}/destinations`, request);
  }
  async updateDestination(
    destId: string,
    updates: Partial<AddDestinationRequest>,
  ): Promise<Destination> {
    return this.client.patch<Destination>(`${this.basePath}/destinations/${destId}`, updates);
  }
  async removeDestination(destId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/destinations/${destId}`);
  }
  async startSimulcast(streamId: string, destinationIds: string[]): Promise<SimulcastSession> {
    return this.client.post<SimulcastSession>(`${this.basePath}/simulcast`, {
      stream_id: streamId,
      destination_ids: destinationIds,
    });
  }
  async stopSimulcast(streamId: string): Promise<SimulcastSession> {
    return this.client.post<SimulcastSession>(`${this.basePath}/simulcast/stop`, {
      stream_id: streamId,
    });
  }
  async getSimulcastStatus(streamId: string): Promise<SimulcastSession> {
    return this.client.get<SimulcastSession>(`${this.basePath}/simulcast/${streamId}`);
  }
  async schedulePost(request: {
    title: string;
    description: string;
    platforms: DestinationType[];
    media_url: string;
    scheduled_at: string;
  }): Promise<ScheduledPost> {
    return this.client.post<ScheduledPost>(`${this.basePath}/posts`, request);
  }
  async listScheduledPosts(params?: PaginationParams): Promise<PaginatedResponse<ScheduledPost>> {
    return this.client.get<PaginatedResponse<ScheduledPost>>(`${this.basePath}/posts`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async cancelScheduledPost(postId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/posts/${postId}`);
  }
  async getDistributionAnalytics(params?: {
    time_range?: string;
    destination_id?: string;
  }): Promise<Record<string, unknown>> {
    return this.client.get<Record<string, unknown>>(`${this.basePath}/analytics`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
}

export function createDistributionAPI(client: WaveClient): DistributionAPI {
  return new DistributionAPI(client);
}
